package/src/nvidiamon.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright (C) CERN, 2020
+
+#include "nvidiamon.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+
+#include "utils.h"
+
+// This is not very nice, but we don't want to add more special runtime
+// flags, pending a proper decision on logging (#106)
+#define NVIDIA_DEBUG 0
+
+// Constructor; uses RAII pattern to be valid after construction
+nvidiamon::nvidiamon()
+    : nvidia_params{prmon::default_nvidia_params},
+      nvidia_stats{},
+      nvidia_average_stats{},
+      nvidia_total_stats{},
+      iterations{0L} {
+  for (const auto& nvidia_param : nvidia_params) {
+    nvidia_stats[nvidia_param] = 0;
+    nvidia_peak_stats[nvidia_param] = 0;
+    nvidia_average_stats[nvidia_param] = 0;
+    nvidia_total_stats[nvidia_param] = 0;
+  }
+
+  // Attempt to execute nvidia-smi
+  // If this works we are valid, but if not then we can't get data
+  valid = test_nvidia_smi();
+}
+
+void nvidiamon::update_stats(const std::vector<pid_t>& pids) {
+  const std::vector<std::string> cmd = {"nvidia-smi", "pmon", "-s",
+                                        "um",         "-c",   "1"};
+
+  for (auto& stat : nvidia_stats) stat.second = 0;
+
+  auto cmd_result = prmon::cmd_pipe_output(cmd);
+  if (cmd_result.first) {
+    // Failed
+    return;
+  }
+
+  if (NVIDIA_DEBUG) {
+    std::cout << "nvidiamon::update_stats got the following output ("
+              << cmd_result.second.size() << "): " << std::endl;
+    int i = 0;
+    for (const auto& s : cmd_result.second) {
+      std::cout << i << " -> " << s << std::endl;
+      ++i;
+    }
+  }
+
+  // Loop over output
+  unsigned int gpu_idx{}, sm{}, mem{}, enc{}, dec{}, fb_mem{};
+  pid_t pid{};
+  std::string cg_type{}, cmd_name{};
+  std::unordered_map<unsigned int, bool>
+      activegpus{};  // Avoid double counting active GPUs
+  for (const auto& s : cmd_result.second) {
+    if (s[0] != '#') continue;
```

Comment:
I think this is supposed to be:

```suggestion
    if (s[0] == '#') continue;
```

as it stands now I'm getting all zeros.

Reply:
Absolutely correct - when I was making this simpler I flipped the logic (or rather *failed* to flip it).

(I couldn't run any test jobs yesterday as the GPUs were busy, so missed this point!)